### ISSUE TITLE
Wrong variable allocation between jumps

### ISSUE THREAD

**Issue by Nekotekina** [asmjit/asmjit#100]
Hi! With the latest revision https://github.com/kobalicek/asmjit/commit/b0dad1af25fb141bf2b20cf29392194886448832 this error occures sometimes in compiler's make() call. Target is W64. Looks like dereferencing nullptr. Access violation reading location 0x0000000000000148.

**Comment by Nekotekina** [comment]
And another problem which may be or may not be related. I managed to debug some crash in generated code. https://gist.github.com/Nekotekina/79c496b9bc00a5205851
It crashed at L4 position with RCX=1 (variable `cpu`), but it's previously allocated to RBX, in L6 block. `mov rcx, rbx` is written after jump, but I suppose it's misplaced or something else is lacking.

**Comment by Nekotekina** [comment]
Also some logs with *** ERROR: Illegal instruction (9). https://gist.github.com/Nekotekina/0ee8f6deb5c36005237b

**Comment by Nekotekina** [comment]
Seems I was able to avoid crashes in loadState by avoiding calling ret() more than once. Maybe it was wrong to have more than one ret() with compiler?

But crashes with wrongly allocated variable still occur.

**Comment by Nekotekina** [comment]
Managed to workaround the problem by allocating some vars to specified x64 registers.

**Comment by kobalicek** [comment]
Hi Ivan, I will check this out tomorrow, I was very busy past week.

**Comment by kobalicek** [comment]
I'm sure all of these Illegal instructions were not processed by the register allocator, I will try to isolate the issue and create some test for it.

**Comment by MichaelRFairhurst** [comment]
I may be seeing the same issue. 

When compiling a function with the shape "if x { ret y }; ret z" I got this as the output ASM via `FileLogger(stdout)`. Fixed it by changing it to "if x { ret y } else ret z".

Broken asm:

```
L0:
push rbx
sub rsp, 32
mov rax, rdi                            ; mov v0, v1
shr rax, 32                             ; shr v0, 32
cmp rax, 65536                          ; cmp v0, 65536
jne L5                                  ; jne L2
xor rax, rax                            ; xor v2, v2
cmp edi, 0                              ; cmp v1, 0
sete rax                                ; sete v2
cmp eax, 0                              ; cmp v2, 0
je L6                                   ; je L3
mov rax, 281474976710656                ; mov v3, 281474976710656
jmp L1
L3:
L4:
mov rcx, rdi                            ; mov v6, v1
sub ecx, 1                              ; sub v6, 1
mov rbx, 281474976710656                ; mov v7, 281474976710656
or rcx, rbx                             ; or v6, v7
mov qword [rsp+16], rdi                 ; [Spill] 
mov rdi, rcx                            ; [Move] 
mov qword [rsp+8], rax                  ; [Save] 
call L0
mov rcx, rax                            ; mov v9, v8
shr rcx, 32                             ; shr v9, 32
cmp rcx, 65536                          ; cmp v9, 65536
jne L2                                  ; jne L2
mov VirtReg<Type=6 Id=261>, VirtReg<Type=6 Id=264>; mov v5, v8
add VirtReg<Type=5 Id=261>, VirtReg<Type=5 Id=257>; add v5, v1
xor VirtReg<Type=6 Id=267>, VirtReg<Type=6 Id=267>; xor v11, v11
mov VirtReg<Type=6 Id=269>, VirtReg<Type=6 Id=257>; mov v13, v1
imul VirtReg<Type=5 Id=269>, VirtReg<Type=5 Id=257>; imul v13, v1
mov VirtReg<Type=6 Id=266>, VirtReg<Type=6 Id=269>; mov v10, v13
idiv VirtReg<Type=5 Id=267>, VirtReg<Type=5 Id=266>, VirtReg<Type=5 Id=257>; idiv v11, v10, v1
mov VirtReg<Type=6 Id=260>, VirtReg<Type=6 Id=261>; mov v4, v5
add VirtReg<Type=5 Id=260>, VirtReg<Type=5 Id=261>; add v4, v5
mov VirtReg<Type=6 Id=270>, 281474976710656; mov v14, 281474976710656
or VirtReg<Type=6 Id=260>, VirtReg<Type=6 Id=270>; or v4, v14
L2:
mov qword [rsp], rax                    ; [Save] 
call 4427282960
L1:
add rsp, 32
pop rbx
ret
L5:
mov rax, qword [rsp]                    ; [Load] 
mov qword [rsp+16], rdi                 ; [Spill] 
short jmp L2
L6:
mov rax, qword [rsp+8]                  ; [Load] 
jmp L3
```

working ASM:

```
L0:
push rbx
sub rsp, 32
mov rax, rdi                            ; mov v0, v1
shr rax, 32                             ; shr v0, 32
cmp rax, 65536                          ; cmp v0, 65536
jne L5                                  ; jne L2
xor rax, rax                            ; xor v2, v2
cmp edi, 0                              ; cmp v1, 0
setne rax                               ; setne v2
cmp eax, 0                              ; cmp v2, 0
je L6                                   ; je L3
mov rax, rdi                            ; mov v5, v1
sub eax, 1                              ; sub v5, 1
mov rbx, 281474976710656                ; mov v6, 281474976710656
or rax, rbx                             ; or v5, v6
mov qword [rsp+16], rdi                 ; [Spill] 
mov rdi, rax                            ; [Move] 
call L0
mov rcx, rax                            ; mov v8, v7
shr rcx, 32                             ; shr v8, 32
cmp rcx, 65536                          ; cmp v8, 65536
jne L7                                  ; jne L2
mov rcx, rax                            ; mov v4, v7
mov rdi, qword [rsp+16]                 ; [Alloc] 
add ecx, edi                            ; add v4, v1
xor rdx, rdx                            ; xor v10, v10
mov rbx, rdi                            ; mov v12, v1
imul ebx, edi                           ; imul v12, v1
mov rax, rbx                            ; mov v9, v12
idiv edx, eax, edi                      ; idiv v10, v9, v1
mov rax, rcx                            ; mov v3, v4
add eax, ecx                            ; add v3, v4
mov rcx, 281474976710656                ; mov v13, 281474976710656
or rax, rcx                             ; or v3, v13
jmp L1
L3:
mov qword [rsp+8], rax                  ; [Spill] 
mov rax, 281474976710656                ; mov v14, 281474976710656
jmp L1
L4:
L2:
mov qword [rsp], rax                    ; [Save] 
call 4516895248
L1:
add rsp, 32
pop rbx
ret
L5:
mov rax, qword [rsp]                    ; [Load] 
short jmp L2
L6:
mov rax, qword [rsp+8]                  ; [Load] 
short jmp L3
L7:
mov rax, qword [rsp]                    ; [Load] 
short jmp L2
```

I too will get illegal instruction errors from this.

**Comment by kobalicek** [comment]
I will try to create a test case based on this, it's a bug

**Comment by MichaelRFairhurst** [comment]
Hopefully the log with comments is useful! If you can't quickly repro I can offer more detail on what its running (virtual register types etc)

**Comment by kobalicek** [comment]
I looked at it but haven't reproduced the issue so far, will look again tomorrow. Btw, what is a signature of the function and what other functions do? Is it something I can rewrite 1:1 or I should simplify? It seems, however, that my simplified case is just passing, but it's kinda too simple I guess:

```x86asm
L0:
sub rsp, 24                             ; 4883EC18
cmp esi, 65535                          ; 81FEFFFF0000            | cmp v0, 65535
jne L2                                  ; 0F85........            | jne L2
cmp esi, edx                            ; 3BF2                    | cmp v0, v1
je L3                                   ; 0F84........            | je L3
mov esi, ecx                            ; 8BF1                    | mov v0, v2
jmp L1                                  ; E9........              | jmp L1
L3:
L4:
mov ecx, edx                            ; 8BCA                    | mov v2, v1
cmp edx, 65535                          ; 81FAFFFF0000            | cmp v1, 65535
jne L5                                  ; 0F85........            | jne L2
mov esi, 128                            ; BE80000000              | mov v0, 128
mov ecx, dword [rsp+8]                  ; 8B4C2408                | [Load] v2
L2:
mov [rdi], esi                          ; 8937                    | mov [dst], v0
mov rdi, qword [rsp]                    ; 488B3C24                | [Load] dst
L1:
add rsp, 24                             ; 4883C418
ret                                     ; C3
L5:
mov ecx, dword [rsp+8]                  ; 8B4C2408                | [Load] v2
short jmp L2                            ; EBEF
[Success] [Alloc] Bug#100.
```

**Comment by kobalicek** [comment]
The best would be to just finish the RA I started in next-wip, but it's more difficult I initially thought

**Comment by MichaelRFairhurst** [comment]
Sorry I haven't gotten back to you on this.

The code I wrote in my higher level language was:

```
fn foo of x is
  if x == 0 do
    ret 0
  end
  ret foo(x - 1) - x * x / x;
end
```

the signature in asmjit is `Function2<uint64_t, uint64_t>`.

Most of the code in there is doing either profiling or typechecking or boxing/unboxing.



I also just ran into maybe the same issue doing a basic recursive fib:

```
fn fib of x is
  if x == 0 do
    ret 0;
  else
    if x == 1 do
      ret 1;
    else
      ret fib(x - 1) + fib(x - 2);
    end
  end
end
```

The generated code self optimizes on the nth call by running the following asm:

```
mov rdi, 140734569204448                ; mov v22, 140734569204448
mov rsi, 140205219643472                ; mov v23, 140205219643472
mov qword [rsp+8], rax                  ; [Save] 
call 4525831264
mov rbx, qword [rsp+24]                 ; [Alloc] 
mov rdi, rbx                            ; mov v25, v2
call rax
```

Except, it wasn't loading `rdi` before `call rax`.

```
mov rdi, 140734569204448                ; mov v22, 140734569204448
mov rsi, 140205219643472                ; mov v23, 140205219643472
mov qword [rsp+8], rax                  ; [Save] 
call 4525831264
call rax
```

until I changed my code to

```
fn fib of x is
  if x == 0 do
    ret 0;
  end
  if x == 1 do
    ret 1;
  end
  ret fib(x - 1) + fib(x - 2);
end
```

and changed my assembler add in some artificial moves:

```
  if (!decl.args.size()) {
    return;
  }

  cc.bind(optimizeLabel);
  X86Gp funcsReg = cc.newUInt64();
  X86Gp declReg = cc.newUInt64();
  X86Gp optimizedAddr = cc.newUInt64();
  cc.mov(funcsReg, (uint64_t) &funcs);
  cc.mov(declReg, (uint64_t) &decl);
  CCFuncCall* recompileCall = cc.call((size_t) &vaiven::optimize, FuncSignature2<uint64_t, uint64_t, uint64_t>());
  recompileCall->setArg(0, funcsReg);
  recompileCall->setArg(1, declReg);
  recompileCall->setRet(0, optimizedAddr);

  // hack bugfix workaround
  vector<X86Gp> argWorkarounds;
  for (int i = 0; i < decl.args.size(); ++i) {
    X86Gp argRegWorkaround = cc.newUInt64();
    cc.mov(argRegWorkaround, argRegs[i]);
    argWorkarounds.push_back(argRegWorkaround);
  }
  CCFuncCall* optimizedCall = cc.call(optimizedAddr, sig);
  for (int i = 0; i < decl.args.size(); ++i) {
    optimizedCall->setArg(i, argWorkarounds[i]);
    //optimizedCall->setArg(i, argRegs[i]);
  }
  X86Gp optimizedRet = cc.newUInt64();
  optimizedCall->setRet(0, optimizedRet);

  cc.ret(optimizedRet);
```

**Comment by kobalicek** [comment]
Just letting you know that I started working on this, but I would like to port the existing register allocator to the next-wip branch as it would be much easier to debug and fix (it uses basic blocks and this should not happen). Not sure how much this can take, but let's see.

**Comment by MichaelRFairhurst** [comment]
Awesome! Great work by the way. I have been really really glad to have your
register allocator available instead of rolling my own. My jit is
performing really well already thanks to your hard work!

**Comment by MichaelRFairhurst** [comment]
It seems like all the issues I was having are solved by #163

**Comment by kobalicek** [comment]
Ok, so I guess I can finally close this one and wait for issues caused by the new RA.
